### PR TITLE
narrow mutex scope when setenv LIBHDFS3_CONF

### PR DESCRIPTION
**Pull request by shuchaome**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
narrow mutex scope when setenv LIBHDFS3_CONF
related issue #35292

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/HDFS/ReadBufferFromHDFS.cpp [hunk 1]
```diff
@@ -22,8 +22,6 @@ ReadBufferFromHDFS::~ReadBufferFromHDFS() = default;
 
 struct ReadBufferFromHDFS::ReadBufferFromHDFSImpl : public BufferWithOwnMemory<SeekableReadBuffer>
 {
-    /// HDFS create/open functions are not thread safe
-    static std::mutex hdfs_init_mutex;
```
Comment (kssenii) [inline src/Storages/HDFS/ReadBufferFromHDFS.cpp:26]: but it is just removing the mutex and leaving not thread-safe functions unprotected

Comment (shuchaome) [reply src/Storages/HDFS/ReadBufferFromHDFS.cpp:26]: The thread unsafe code I found in ReadBufferFromHDFS is setenv/getenv LIBHDFS3_CONF and i think the mutex is for it.（Any other code unsafe？）
Using std::call_once for setenv is more efficient than the current static mutex. (getenv no need to protect)

Comment (shuchaome) [reply src/Storages/HDFS/ReadBufferFromHDFS.cpp:26]: The thread unsafe code I found in ReadBufferFromHDFS is setenv/getenv LIBHDFS3_CONF and i think the mutex is for it.（Any other code unsafe？）
Using std::call_once for setenv is more efficient than the current static mutex. (getenv no need to protect)

Comment (shuchaome) [reply src/Storages/HDFS/ReadBufferFromHDFS.cpp:26]: @alesapin  Could you help to review？ The mutex was added in #15675

Comment (kssenii) [reply src/Storages/HDFS/ReadBufferFromHDFS.cpp:26]: > The thread-unsafe code I found in ReadBufferFromHDFS is setenv/getenv LIBHDFS3_CONF and i think the mutex is for it.（Any other unsafe？）

As the comment states: it is hdfs functions which are not thread safe, in libhdfs3 library

**Review by kssenii — APPROVED** [review]
ok, given that integration tests with tsan passed, this change is ok.

**Comment by kssenii** [comment]
But need to merge with master, because I see `"This branch has conflicts that must be resolved"`

**Comment by shuchaome** [comment]
Now it is ready for merged.

**Pull request merged**